File: examples/python/ticrate.py
#!/usr/bin/env python3

from __future__ import print_function
from multiprocessing import Process
from random import choice
import vizdoom as vzd
from argparse import ArgumentParser

DEFAULT_CONFIG = "../../scenarios/basic.cfg"


def play(config_file, ticrate=35):
    game = vzd.DoomGame()

    game.load_config(config_file)
    game.set_mode(vzd.Mode.ASYNC_PLAYER)

    game.set_ticrate(ticrate)

    game.init()

    actions = [[True, False, False],
               [False, True, False],
               [False, False, True]]
    episodes = 10

    for i in range(episodes):
        game.new_episode()

        while not game.is_episode_finished():
            game.make_action(choice(actions))

    game.close()


if __name__ == '__main__':
    if __name__ == "__main__":
        parser = ArgumentParser("ViZDoom example showing how to change the ticrate for asynchronous mode.")
        parser.add_argument(dest="config",
                            default=DEFAULT_CONFIG,
                            nargs="?",
                            help="Path to the configuration file of the scenario."
                                 " Please see "
                                 "../../scenarios/*cfg for more scenarios.")
        parser.add_argument("-t", "--ticrates",
                            default=[17,35,70],
                            nargs="+",
                            help="List of ticrates to show.")
        args = parser.parse_args()

        processes= []
        for ticrate in args.ticrates:
            p = Process(target=play, args=[args.config, ticrate])
            p.start()
            processes.append(p)

        for p in processes:
            p.join()

File: examples/lua/ticrate.lua
#!/usr/bin/env th

local threads = require 'threads'

local noOfPlayers = 2
local episodes = 10

local pool = threads.Threads(

    -- spawned threads
    noOfPlayers,

    -- Set-up the environment on each thread
    function()
        -- for some reason 'require' has an unexpected behavior when called
        -- from 'threads' to execute a module outside this director.
        require "vizdoom"

        actions = {
            [1] = torch.IntTensor({1,0,0}),
            [2] = torch.IntTensor({0,1,0}),
            [3] = torch.IntTensor({0,0,1})
        }
    end,

    function()
        function player1()
            -- Create DoomGame instance.
            game = vizdoom.DoomGame()

            -- Config
            game:loadConfig("../config/basic.cfg")
            game:setMode(vizdoom.Mode.ASYNC_PLAYER)

            -- Default Doom's ticrate is 35 per second,
            -- so this one will work 2 times faster.
            game:setTicrate(70)

            game:init()

            for i = 1, episodes do
                game:newEpisode()
                print("New episode: ", __threadid)

                while not game:isEpisodeFinished() do
                    -- Make a random action
                    local action = actions[torch.random(#actions)]
                    reward = game:makeAction(action)
                end
            end
            game:close()
        end

        function player2()
            -- Create DoomGame instance.
            game = vizdoom.DoomGame()

            -- Config
            game:loadConfig("../config/basic.cfg")
            game:setMode(vizdoom.Mode.ASYNC_PLAYER)

            -- And this one will work 2 times slower.
            game:setTicrate(17)

            game:init()

            for i = 1, episodes do
                game:newEpisode()
                print("New episode: ", __threadid)

                while not game:isEpisodeFinished() do
                    -- Make a random action
                    local action = actions[torch.random(#actions)]
                    reward = game:makeAction(action)
                end
            end
            game:close()
        end
    end
)

pool:specific(true)

pool:addjob(
    1,
    function()
        player1()
    end
)
pool:addjob(
    2,
    function()
        player2()
    end
)

pool:synchronize()
pool:terminate()
print("Done")
